extract Ant modules, replace commons.io, extractTCasesCommand

tkruse opened the pull request:
Solves #15, #17, #18 

I created 2 new maven modules, tcases-core and tcases-ant.
tcases-core contains most of previous tcases-lib.

tcases-lib only contains Tcases.java, and Reducer.Java (also it directly depends on logback).
I kept the name tcases-lib so that previous users of tcases may only upgrade a version number and still get the same classes (except for the ant task). But renaming of maven modules may also be fine if mentioned in release notes.

Also I split `Tcases.java` into `Tcases.java` and `TCases.Core.java`, with an additional slightly ugly split of `TCases.Options`  into `TCases.Options` and `TCasesCore.Options`.

I removed the compile dependency to commons.io by using plain Java code instead of IOUtils.

In Maven I moved dependency version and scopes to the root pom for easier management, and I replaced direct dependencies to logback to the slf4j wrapper that the code already uses. So tcases-core does not directly depend on logback anymore, slf4J makes sure library users can use any logging system.

I see that you actually use the Tcases xml files in Junit to test the tcases generators. Because of that (and because of Maven limitations https://stackoverflow.com/questions/16948748), it is practically impossible to split the io packages from the non-io combinatorial code (unless you would consider migrating to Gradle, which is not trivial).

So things for you to decide:
* Naming of the maven modules
* Naming of modified Classes TCases, TCases.Options, TCasesCore, TCasesCore.Options
* Review of deprecated TCases.getTests methods

Other than that, to the best of my knowlegde I did not change any java code. Github should show no difference for moved files

For comparison: Dependency:tree before and after:

    # Before

    org.cornutum.tcases:tcases-lib:jar:1.5.5-SNAPSHOT
     +- commons-io:commons-io:jar:2.4:compile
     +- org.apache.commons:commons-lang3:jar:3.3.1:compile
     +- org.apache.commons:commons-collections4:jar:4.1:compile
     +- ch.qos.logback:logback-classic:jar:1.1.1:compile
     |  +- ch.qos.logback:logback-core:jar:1.1.1:compile
     |  \- org.slf4j:slf4j-api:jar:1.7.6:compile
    \- org.apache.ant:ant:jar:1.8.2:compile
        \- org.apache.ant:ant-launcher:jar:1.8.2:compile

    # After

    org.cornutum.tcases:tcases-core:jar:1.5.5-SNAPSHOT
    +- org.apache.commons:commons-lang3:jar:3.3.1:compile
    +- org.apache.commons:commons-collections4:jar:4.1:compile
    +- org.slf4j:slf4j-api:jar:1.7.6:compile

tkruse commented on pom.xml:
```diff
@@ -177,7 +177,51 @@
+      <dependency>
+        <groupId>junit</groupId>
+        <artifactId>junit</artifactId>
+        <version>4.12</version>
```
sorry, this version upgrade slipped in (was 4.11), though I think it might be ok. I did not update the other versions.

tkruse commented on tcases-ant/pom.xml:
```diff
@@ -0,0 +1,112 @@
+    <maven.build.timestamp.format>yyyy-MM-dd</maven.build.timestamp.format>
+  </properties>
+
+  <build>
```
I copy/pasted the pom.xml files for the new modules. So the lines below are not currently useful, but at least they are consistent.

tkruse commented on tcases-lib/pom.xml:
```diff
@@ -84,41 +84,48 @@
+      <groupId>org.cornutum.tcases</groupId>
+      <artifactId>tcases-core</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
```
See https://stackoverflow.com/questions/14722873 for an explanation

tkruse commented on tcases-lib/src/main/java/org/cornutum/tcases/Tcases.java:
```diff
@@ -1185,141 +847,58 @@ else if( options.getTransformType() == Options.TransformType.HTML)
+    /*
+     * @deprecated Use TcasesCore.getTests
+     */
+    @Deprecated()
```
all these deprecated methods below are of course not strictly required, just provided as a courtesy to clients

kerrykimbrough commented on tcases-core/src/main/java/org/cornutum/tcases/io/SystemTestHtmlWriter.java:
```diff
@@ -13,13 +13,9 @@
-import java.io.IOException;
-import java.io.OutputStream;
-import java.io.Writer;
+import java.io.*;
```
Importing everything from a large JDK package can be problematic. I recommend importing each class actually used individually.

tkruse replied:
fixed

kerrykimbrough commented on tcases-core/src/main/java/org/cornutum/tcases/io/SystemTestHtmlWriter.java:
```diff
@@ -311,21 +307,19 @@ protected void writeDefaultStyle()
-      for( styleLines = IOUtils.lineIterator( getClass().getResourceAsStream( "system-test.css"), "UTF-8");
-           styleLines.hasNext();
-           xmlWriter_.println( styleLines.next()));
+      BufferedReader in = new BufferedReader( new InputStreamReader( getClass().getResourceAsStream( "system-test.css"), "UTF-8"));
```
Oops, classic defect here: does not ensure input stream is always closed, even after an exception. For example, the original `finally` clause is missing. 

But instead of `IOUtils.closeQuietly`, should now use standard Java try-with-resources.

tkruse replied:
fixed

kerrykimbrough commented on tcases-core/src/main/java/org/cornutum/tcases/io/SystemTestHtmlWriter.java:
```diff
@@ -339,21 +333,21 @@ protected void writeDefaultScript()
-      for( scriptLines = IOUtils.lineIterator( getClass().getResourceAsStream( "system-test.js"), "UTF-8");
-           scriptLines.hasNext();
-           xmlWriter_.println( scriptLines.next()));
+      BufferedReader in = new BufferedReader( new InputStreamReader( getClass().getResourceAsStream( "system-test.js"), "UTF-8"));
```
See previous comment

tkruse replied:
fixed

kerrykimbrough commented:
I think there's a better way to structure these changes.  The main goal is to decouple the core from command line support. There already exists a module=tcases-shell for command line support. Currently, tcases-shell.jar is empty, but now that's where command line classes like Tcases, Reducer, etc. should be located. In which case, the existing tcases-lib module becomes the "core" and no additional module is needed. (Although "tcases-core" is a better name, so I don't object to renaming it.)

kerrykimbrough commented:
Note that there is no need for the Tcases.Options class. The Tcases class simply needs a method to convert command line args into a TcasesCore.Options class.

tkruse commented on tcases-lib/src/main/java/org/cornutum/tcases/TcasesCommand.java:
```diff
@@ -301,39 +298,29 @@
    *
    */
-  public static class Options
+  public static class OptionFactory
```
Since there were multiple methods Involved, I kept using a class.

tkruse commented:
commit 96d3bc6  moves most classes back to tcases-lib. I tried making tcases-ant only depend on tcases-core, but that needed the run() method of TCases Command (and getProjectName() and getVersion()), and then I noticed that Tcases has zero testcoverage other than the tests of Tcases command, and things just got progressively uglier.

With that commit, the situation is as following compared to master:

* A new tcases-ant module exists with the ant task and dependency
* Tcases class has been split into Tcases and TcasesCommand, the latter had OptionFactory with parts of previous Tcases.Options
* commons.io/IOUtils methods used in main code have been replaced with pure java
* the logback dependency is only a runtime dependency in tcases-shell, else slf4j is used.

The github diff should reflect the changeset having become much smaller. I could also merge back TCases and TCasesCommand into a single class, that's up to you. Moving the CLI stuff to tcases-shell might make sense, but with the test coverage situation being what it is, I think it would not be an improvement.

tkruse commented on tcases-lib/src/main/java/org/cornutum/tcases/util/FilenameUtils.java:
```diff
@@ -0,0 +1,19 @@
+public class FilenameUtils
+  {
+
+  public static String getBaseName( String filepath)
```
The code is heavily inspired by commons.io IOUtils. I added a testcase too.

kerrykimbrough replied:
Why is this better than using the original commons-io dependency?

tkruse replied:
So my goal was to remove the common-io compile dependency to use tcases as a library with minimal transitive dependencies (and thus minimal potential for conflicts with other libraries). commons.io seemed to be used so little that removing it seemed feasible at a low price. But you can disagree saying that it is a benign dependency and move it back in, since the method and classnames used are the same, it's just a matter of deleting 3 Java files and reverting the imports in 3 places or so.

tkruse replied:
else this getBaseName and the commons.io.FileUtils.getBaseName methods should be equivalent.

kerrykimbrough replied:
Agreed, the need for common-io is negligible, so I'm OK with replacing it.

kerrykimbrough commented on tcases-lib/src/main/java/org/cornutum/tcases/TcasesCommand.java:
```diff
@@ -0,0 +1,582 @@
+public class TcasesCommand extends Tcases
+  {
+  /**
+   * Represents a set of command line options.
```
Should change now to:

```
Returns the {@link Options} represented by a set of command line arguments.
```

kerrykimbrough commented on tcases-lib/src/main/java/org/cornutum/tcases/TcasesCommand.java:
```diff
@@ -0,0 +1,582 @@
+   * <P/> 
+   *
+   */
+  public static class OptionFactory
```
A better name would be "OptionsFactory".

kerrykimbrough commented on tcases-lib/src/main/java/org/cornutum/tcases/TcasesCommand.java:
```diff
@@ -0,0 +1,582 @@
+   * Writes an XML document describing given test case definitions to the given output stream.
+   */
+  @SuppressWarnings("resource")
+  public static void writeTests( SystemTestDef testDef, OutputStream outputStream)
```
For consistency, this method ought to be defined by the `Tcases` class.

kerrykimbrough commented:
Need to make the following changes to the docs:

```
File: tcases-shell/docs/Tcases-Guide.htm
@@ -279,7 +279,7 @@
       <P>
         For details about the interface to the <SPAN class="code">tcases</SPAN> command (and
         the <SPAN class="code">tcases.bat</SPAN> command, too), see the Javadoc for
-        the <A href="api/org/cornutum/tcases/Tcases.Options.html"><SPAN class="code">Tcases.Options</SPAN></A>
+        the <A href="api/org/cornutum/tcases/TcasesCommand.OptionsFactory.html"><SPAN class="code">TcasesCommand.OptionsFactory</SPAN></A>
         class.  To get help at the command line, run <SPAN class="code">tcases -help</SPAN>.
       </P>
 
@@ -1604,7 +1604,7 @@
       </P>
       <P>
         Of course, you can use various options for the <SPAN class="code">tcases</SPAN> command to customize this default
-        pattern. For details, see the <A href="api/org/cornutum/tcases/Tcases.Options.html"><SPAN class="code">Tcases.Options</SPAN></A> class,
+        pattern. For details, see the <A href="api/org/cornutum/tcases/TcasesCommand.OptionsFactory.html"><SPAN class="code">TcasesCommand.OptionsFactory</SPAN></A> class,
         or run <SPAN class="code">tcases -help</SPAN>.
       </P>
 
@@ -2456,7 +2456,7 @@ public class Examples { <BR/>
 
         <UL>
           <LI> The <SPAN class="code">find</SPAN> command <A href="examples/find-Input.xml">example</A></LI><P/>
-          <LI> The <SPAN class="code">tcases</SPAN> <A href="api/org/cornutum/tcases/Tcases.Options.html">command line</A></LI><P/>
+          <LI> The <SPAN class="code">tcases</SPAN> <A href="api/org/cornutum/tcases/TcasesCommand.OptionsFactory.html">command line</A></LI><P/>
           <LI> Document schemas <P/>
             <UL>
               <LI> The <A href="system-input.xsd">system input definition document schema</A> </LI><P/>
```

tkruse commented:
I also changed the html doc as you described.

If you see more small changes to fix, it might be faster if you fix those yourself.

You can easily get the PR branch like this:
https://github.com/Cornutum/tcases/pull/19#issuecomment-399699103

```
git fetch origin pull/19/head:feature/extract
git checkout feature/extract
```

If you push that branch to you repo, I can make PRs against that branch with just changes to add to the branch before merging the branch to Master.

kerrykimbrough commented:
> You can easily get the PR branch like this...

Thanks, I never learned about that feature before. 

Or, if you're willing, it might be simpler to allow me to make changes directly to your PR branch. Here's how:  https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

kerrykimbrough commented:
> Shouldn't the following methods be moved from Tcases to TcasesCommand?

Oops! References to relocated functions must be changed from `Tcases.*` to `TcasesCommand.*`.

tkruse commented:
I did not know about the feature to allow maintainer modifications of PRs. In any case, that was already enabled, so I guess you could already do that.

I added a commit to fix method invocations for relocated methods

tkruse commented:
Sorry, I only now started looking in detail at the Options fields, and I realzied all but 2 are only relevant to TCasesCommand. So I moved Options in it's original form to TCasesCommand, and changed the interface of Tcases.getTests() to take the only fields from options it uses, Long seed and Integer defaultTupleSize. Those could be wrapped into a much smaller Options bean to keep the interface stable when more options become relevant to TCases. I think overall this reduces the PRs diff more (no more OptionsFactory), so that should be good.

kerrykimbrough commented:
Good point. I agree that options should be aggregated in a bean at all
levels. These beans should have names that better convey their purpose. For
example, CommandOptions which extends RunOptions.

On Sun, Jun 24, 2018, 10:36 PM Thibault Kruse <notifications@github.com>
wrote:

> Sorry, I only now started looking in detail at the Options fields, and I
> realzied all but 2 are only relevant to TCasesCommand. So I moved Options
> in it's original form to TCasesCommand, and changed the interface of
> Tcases.getTests() to take the only fields from options it uses, Long seed
> and Integer defaultTupleSize. Those could be wrapped into a much smaller
> Options bean to keep the interface stable when more options become relevant
> to TCases. I think overall this reduces the PRs diff more (no more
> OptionsFactory), so that should be good.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Cornutum/tcases/pull/19#issuecomment-399821426>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGRe-faejoQtGFk_UaPS8mujy0_Ul3IFks5uAFrWgaJpZM4UzNKA>
> .
>

kerrykimbrough commented:
~~For example, CommandOptions which extends RunOptions.~~

FYI, I've committed this change, but in a slightly different way.

tkruse commented:
GeneratorOptions looks ok to me. So you will merge this PR as it is now? I have no plans to add more to it.

kerrykimbrough commented:
I'm working on the one small clean up related to the ant module, after which I will complete the merge.
